### PR TITLE
[WIP][Model] Add support for multiple audio chunks/audio URLs

### PR DESCRIPTION
**Pull request by petersalas**
This adds:
 - Support for `NestedTensors` of arbitrary depth to represent multimodal inputs -- model implementations that support multiple multimodal inputs can return lists from their input mapper to include a dimension that represents the number of inputs in the prompt.
 - Support for multiple audio inputs per prompt in `UltravoxModel`
 - Support for multiple `"audio_url"`/`"image_url"` items in the OpenAI API server (up to the limits provided at initialization)

This depends on #7230 (for changes to `repeat_and_pad_placeholder_tokens`) which in turn is blocked on #7783.

I intend to split this into three PRs (NestedTensors, Ultravox, API server) but I'd love to get an assessment from @DarkLight1337 and @ywang96 on the overall approach.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/entrypoints/chat_utils.py [hunk 1]
```diff
@@ -143,13 +194,37 @@ def _mm_token_str(model_config: ModelConfig, tokenizer: AnyTokenizer,
...
-def _get_full_multimodal_text_prompt(placeholder_token_str: str,
+def _get_full_multimodal_text_prompt(placeholders: List[str],
                                      text_prompt: str) -> str:
     """Combine multimodal prompts for a multimodal language model"""
 
+    # Look through the text prompt so that we don't add placeholders for
+    # items that already have them.
+    prompt_fragments = [text_prompt]
+    missing_placeholders: List[str] = []
+    for placeholder in placeholders:
```
Comment (petersalas) [inline vllm/entrypoints/chat_utils.py:205]: This is pretty convoluted as it's a full generalization of the check that was there before -- happy to do something much simpler, I just wasn't sure which "placeholder is already present in the prompt" scenarios are important.

Comment (petersalas) [reply vllm/entrypoints/chat_utils.py:205]: (Also it needs tests if it remains in it's current form :))

File: vllm/multimodal/base.py [hunk 2]
```diff
@@ -54,26 +52,32 @@ class MultiModalInputs(_MultiModalInputsBase):
...
+        concatenated = [
+            MultiModalInputs._try_concat(t, is_nested=True)
+            for t in nested_tensors
+        ]
+        if any(isinstance(t, list) for t in concatenated):
+            return concatenated
+
+        tensors_ = cast(List[torch.Tensor], concatenated)
+        return torch.stack(tensors_) if all(t.shape == tensors_[0].shape
+                                            for t in tensors_) else tensors_
```
Comment (DarkLight1337) [inline vllm/multimodal/base.py:80]: Thanks for implementing my suggestion from the RFC! Would be great to add some tests for this.

File: tests/models/test_ultravox.py [hunk 3]
```diff
@@ -70,7 +72,7 @@ def vllm_to_hf_output(vllm_output: Tuple[List[int], str,
 def run_test(
```
Comment (DarkLight1337) [inline tests/models/test_ultravox.py:72]: Let's separate the test for single-input (without list) and multi-input (with list)

File: vllm/entrypoints/chat_utils.py [hunk 4]
```diff
@@ -80,10 +80,61 @@ class ConversationMessage(TypedDict):
...
+                        if isinstance(existing_item, list):
+                            result_list = existing_item
+                        else:
+                            result_list = [existing_item]
+                            mm_data[mm_key] = result_list  # type: ignore
+
+                        if isinstance(mm_item, list):
+                            result_list.extend(mm_item)
+                        else:
+                            result_list.append(mm_item)
```
Comment (DarkLight1337) [inline vllm/entrypoints/chat_utils.py:133]: Looks quite complicated. Can we simplify the code here using a `defaultdict`? We allow `Mapping[str, object]` in `MultiModalDataDict` so this shouldn't cause any type errors.

File: vllm/entrypoints/chat_utils.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/entrypoints/chat_utils.py:205]: I think this is fine, but definitely needs some examples to better make sense of it.

**Review by DarkLight1337 — COMMENTED** [review]
Left some initial comments!

File: vllm/multimodal/base.py (hunk 2 above)
Comment (petersalas) [reply vllm/multimodal/base.py:80]: Thank you for laying out the plan!

**Comment by petersalas** [comment]
Closing in favor of #7902 and #7923 (and another PR to follow for Ultravox).
